### ISSUE TITLE
aead: In place API but with different buffers

### ISSUE THREAD

**Issue by ia0** [RustCrypto/traits#1311]
Would it make sense to add the following methods to `AeadInPlace` (similarly for `AeadMutInPlace`) with default implementations based on the existing methods?

```rust
fn encrypt_mut(&self, nonce: &Nonce, aad: &[u8], plain: &[u8], cipher_tag: &mut [u8]) -> Result<()> { ... }
fn encrypt_mut_detached(&self, nonce: &Nonce, aad: &[u8], plain: &[u8], cipher: &mut [u8]) -> Result<Tag> { ... }

fn decrypt_mut(&self, nonce: &Nonce, aad: &[u8], cipher_tag: &[u8], plain: &mut [u8]) -> Result<()> { ... }
fn decrypt_mut_detached(&self, nonce: &Nonce, aad: &[u8], cipher: &[u8], tag: &Tag, plain: &mut [u8]) -> Result<()> { ... }
```

This would add support (without extra copies or allocation) for implementations that don't support aliasing input and output. They would implement the `_mut` version and use a copy to implement the `_in_place` version using the `_mut` version.

**Comment by tarcieri** [comment]
We could potentially add something like this, but I'll note the API surface is already pretty complicated.

I'm not sure it makes sense to add these to `AeadInPlace`: they're not in-place operations. But we could potentially add yet another trait with a blanket impl for types which impl `AeadInPlace`.

**Comment by ia0** [comment]
I agree the current API is complicated but may I suggest to make it more complicated to possibly make it simpler:
- There are 6 "implementer" traits: one per type of implementation (in-place, 2 buffers, or alloc on one side and stateful or stateless on the other, so 6 traits). Those traits would only be used by implementers.
- There are 6 "user" traits: one per type of usage (similar to the implementer traits but with a different name). Those traits would only be used by users.
- A blanket impl from each implementer trait to each user trait when possible. Note that some user traits may not be implementable by any implementer trait.

**Comment by tarcieri** [comment]
Pretty much all of our AEADs impl `AeadCore`+`AeadInPlace`, and there's already a blanket impl of `Aead` for `AeadCore`.

The `*Mut` traits are pretty much there to support hardware accelerators and we don't otherwise impl stateful AEADs.

We don't really have a distinction of "implementer" versus "user" traits: each trait is independently useful to end users.

Such a distinction exists in the `digest` crate e.g. `digest::core_api` but I'm not sure that would be something we want to carry over here.

**Comment by ia0** [comment]
Yes I understand the current design, I'm suggesting to fix its limitations, in particular the fact that blanket impls force extra copy when implemented and used interfaces differ. Here's a concrete example:
- Let's call AeadNoAlloc the 2 buffers version.
- Let's assume we have a blanket impl from AeadInPlace to AeadNoAlloc.
- Let's assume the hardware only provides a 2 buffers API.
- You either have to only implement AeadNoAlloc and prevent users from using AeadInPlace. Or worse you have to implement AeadInPlace with allocation and copy, then when the user wants to use AeadNoAlloc they pay this price plus another copy from the blanket impl.

By dissociating implementer and user interfaces you provide as much choice to the user as possible, paying only the minimum interface mismatch cost.

I can see how it may be more complicated by having much more traits, but the design is also easier to understand: if you're a user, just use the trait (possibly multiple) you want from those, if you're an implementer, just implement the trait (and only one) you want from those.

No worries if that's too much complexity. I'm anyway currently adding the missing traits myself. It's just that it could be useful for others too.

**Comment by tarcieri** [comment]
To back up, what you want isn't an in-place API, but one which has an immutable reference to the source and a mutable reference to the destination buffer. So, it doesn't make sense to put such methods on the `AeadInPlace` trait, because they aren't in-place methods (which definitionally operate over the same input/output buffer).

So what you're asking for only makes sense as a new trait.

>  I'm suggesting to fix its limitations, in particular the fact that blanket impls force extra copy when implemented and used interfaces differ. 

I don't think this scenario would happen. The blanket impl would exist for types which impl `AeadInPlace`, but for backends which don't support in-place operation and natively prefer separate source/destination buffers, you wouldn't impl `AeadInPlace`, only the new traits.

Note that [`AeadMut`](https://docs.rs/aead/latest/aead/trait.AeadMut.html) already exists, so the `encrypt_mut*` and `decrypt_mut*` names don't make sense.

What might make sense is adding methods to the [`Aead`](https://docs.rs/aead/latest/aead/trait.Aead.html) trait, and naming them something like `encrypt_into` and `decrypt_into`. That would probably be a breaking change though.

**Comment by ia0** [comment]
> what you want isn't an in-place API

Indeed, the naming is wrong. I called it like that because I also consider in-place the version of an API that takes an output buffer compared to one that allocates and returns a new buffer, because it essentially operates in-place instead of creating a new-place. But let's stick to the current naming which makes more sense when there's both an input and an output buffer.

> I don't think this scenario would happen. The blanket impl would exist for types which impl AeadInPlace, but for backends which don't support in-place operation and natively prefer separate source/destination buffers, you wouldn't impl AeadInPlace, only the new traits.

Yes, that's also an option. But then, you put the weight to handle this impedance on the user. Because now users can't just use an in-place API and be portable over all implementations. They need somehow to check whether the implementation supports an in-place API and if not, use another API. This argument is symmetric, I used in-place, but the user should be able to use any API A regardless of whether the implementation implements A, B, or C, since they're all equivalent modulo possible copy and/or allocation.

> That would probably be a breaking change though.

Yes, I think any solution will be a breaking change. Only adding a new independent trait would not be a breaking change. That would not be the best user experience, but that would be fine for me. I'll implement my own impedance matching traits on top.

**Comment by newpavlov** [comment]
We have the [`inout`](https://docs.rs/inout) crate specifically to generalize over in-place and buffer-to-buffer modes of operation. The cipher crates already implemented on top of it. Currently `inout` supports only cases when input and output data is of equal length, so it would be limited to AEADs based on stream ciphers used in detached mode.

**Comment by tarcieri** [comment]
Adding `inout` support to the detached modes sounds like a good idea to me

**Comment by MarkAtwood** [comment]
Related: #2360, filed by the wolfCrypt backend ([wolfSSL/wolfssl-rs](https://github.com/wolfSSL/wolfssl-rs)), documents a concrete real-world case of this problem. wolfCrypt's one-shot `wc_ChaCha20Poly1305_Encrypt` requires separate source and destination pointers; adapting it to `AeadInPlace` requires a heap allocation per call or switching to a streaming API. The workaround used and its limitations are documented at [wolfcrypt/src/aead.rs:198-207](https://github.com/wolfSSL/wolfssl-rs/blob/0f3af10/wolfcrypt/src/aead.rs#L198-L207).
